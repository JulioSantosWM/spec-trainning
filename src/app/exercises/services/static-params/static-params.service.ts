import { Inject, Injectable } from "@angular/core";
import { STATIC_PARAMS } from "../../mocks/static-params.mock";
import { StaticParams } from "../../models/static-params.model";
import { DisplayName } from "../../enums/display-name.enum";
import { PrimaryContact } from "../../enums/primary-contact.enum";

@Injectable()
export class StaticParamsService {

  constructor(@Inject(STATIC_PARAMS) private readonly staticParams: StaticParams) {}

  get primaryContact(): PrimaryContact {
    if (this.staticParams?.primaryContactType !== undefined && Object.values(PrimaryContact).includes(this.staticParams.primaryContactType)) {
      return this.staticParams.primaryContactType;
    }
    throw new Error('No primary contact set in StaticParams.')
  }

  get displayName(): DisplayName {
    if (this.staticParams?.displayNameType !== undefined && Object.values(DisplayName).includes(this.staticParams.displayNameType)) {
      return this.staticParams.displayNameType;
    }
    return DisplayName.FullName;
  }

  get expandedDetails(): boolean {
    if (this.staticParams?.expandedDetails !== undefined) {
      return this.staticParams.expandedDetails;
    }
    return false;
  }
}