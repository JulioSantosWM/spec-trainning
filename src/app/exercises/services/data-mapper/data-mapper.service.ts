import { Injectable } from "@angular/core";
import { UserDto } from "../../models/user.dto";
import { User } from "../../models/user.model";
import { CreditCardDto } from "../../models/credit-card.dto";
import { StaticParamsService } from "../static-params/static-params.service";
import { DisplayName } from "../../enums/display-name.enum";
import { PrimaryContact } from "../../enums/primary-contact.enum";

@Injectable()
export class DataMapperService {

  constructor(private staticParamsService: StaticParamsService) {}

  toUser(userDto: UserDto): User {
    const user: User = {
      id: userDto.id,
      uid: userDto.uid,
      username: userDto.username,
      name: this.getName(userDto),
      date_of_birth: new Date(userDto.date_of_birth),
      primary_contact: this.getPrimaryContact(userDto),
      avatar: userDto.avatar,
      address: {
        ...userDto.address
      }
    }
    return user;
  }

  toCreditCard(creditCardDto: CreditCardDto): CreditCardDto {
    return creditCardDto;
  }

  private getName(userDto: UserDto): string {
    switch (this.staticParamsService.displayName) {
      case DisplayName.FirstName: 
        return userDto.first_name;
      case DisplayName.LastName:
        return userDto.last_name;
      default:
        return userDto.first_name + ' ' + userDto.last_name;
    }
  }

  private getPrimaryContact(userDto: UserDto): string {
    if (this.staticParamsService.primaryContact === PrimaryContact.Email) {
      return userDto.email;
    }
    return userDto.phone_number;
  }
}