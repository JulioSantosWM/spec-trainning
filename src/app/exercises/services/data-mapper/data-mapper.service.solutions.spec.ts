import { MockBuilder, MockInstance, MockRender } from "ng-mocks";
import { DataMapperService } from "./data-mapper.service"
import { StaticParamsService } from "../static-params/static-params.service";
import { createMock } from "@golevelup/ts-jest";
import { UserDto } from "../../models/user.dto";
import { DisplayName } from "../../enums/display-name.enum";
import { PrimaryContact } from "../../enums/primary-contact.enum";
import { CreditCardDto } from "../../models/credit-card.dto";

describe('DataMapperService', () => {
  let service: DataMapperService;

  MockInstance.scope();

  beforeEach(() => {
    return MockBuilder(DataMapperService)
      .mock(StaticParamsService)
  })

  function setup(): void {
    service = MockRender(DataMapperService).point.componentInstance;
  }

  describe('when mapping to a user', () => {
    const mock = createMock<UserDto>({
      uid: 'foo',
      first_name: 'bar',
      last_name: 'baz',
      phone_number: 'qux',
      email: 'quux'
    });

  
    MockInstance.scope();

    it('should map a user according to the default mapping', () => {
      MockInstance(StaticParamsService, 'displayName', DisplayName.FullName);
      MockInstance(StaticParamsService, 'primaryContact', PrimaryContact.Phone);

      setup();

      const result = service.toUser(mock);

      expect(result).toMatchSnapshot();
    })
  
    it('should map a user with the name as first_name', () => {
      MockInstance(StaticParamsService, 'displayName', DisplayName.FirstName);
      MockInstance(StaticParamsService, 'primaryContact', PrimaryContact.Phone);

      setup();

      const result = service.toUser(mock);

      expect(result).toMatchSnapshot();
    })
    it('should map a user with the name as last_name', () => {
      MockInstance(StaticParamsService, 'displayName', DisplayName.LastName);
      MockInstance(StaticParamsService, 'primaryContact', PrimaryContact.Phone);

      setup();

      const result = service.toUser(mock);

      expect(result).toMatchSnapshot();
    })
    it('should map a user with the primary contact as email', () => {
      MockInstance(StaticParamsService, 'displayName', DisplayName.FullName);
      MockInstance(StaticParamsService, 'primaryContact', PrimaryContact.Email);

      setup();

      const result = service.toUser(mock);

      expect(result).toMatchSnapshot();
    })
  })

  describe('when mapping to a credit card', () => {
    const mock = createMock<CreditCardDto>({ uid: 'foo' });
  
    MockInstance.scope();

    it('should map to a credit card', () => {
      setup();

      const result = service.toCreditCard(mock)

      expect(result).toEqual(result);
    })
  })
})