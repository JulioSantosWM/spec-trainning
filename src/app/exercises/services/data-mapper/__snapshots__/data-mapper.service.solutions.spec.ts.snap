// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataMapperService when mapping to a user should map a user according to the default mapping 1`] = `
{
  "address": [MockFunction],
  "avatar": [MockFunction],
  "date_of_birth": Date { NaN },
  "id": [MockFunction],
  "name": "bar baz",
  "primary_contact": "qux",
  "uid": "foo",
  "username": [MockFunction],
}
`;

exports[`DataMapperService when mapping to a user should map a user with the name as first_name 1`] = `
{
  "address": [MockFunction],
  "avatar": [MockFunction],
  "date_of_birth": Date { NaN },
  "id": [MockFunction],
  "name": "bar",
  "primary_contact": "qux",
  "uid": "foo",
  "username": [MockFunction],
}
`;

exports[`DataMapperService when mapping to a user should map a user with the name as last_name 1`] = `
{
  "address": [MockFunction],
  "avatar": [MockFunction],
  "date_of_birth": Date { NaN },
  "id": [MockFunction],
  "name": "baz",
  "primary_contact": "qux",
  "uid": "foo",
  "username": [MockFunction],
}
`;

exports[`DataMapperService when mapping to a user should map a user with the primary contact as email 1`] = `
{
  "address": [MockFunction],
  "avatar": [MockFunction],
  "date_of_birth": Date { NaN },
  "id": [MockFunction],
  "name": "bar baz",
  "primary_contact": "quux",
  "uid": "foo",
  "username": [MockFunction],
}
`;
